cpp/src/arrow/compute/kernels/scalar_if_else.cc

```diff
@@ -0,0 +1,740 @@
+      if (right_data) {
+        // out_buf = ~cond
+        ARROW_ASSIGN_OR_RAISE(out_buf, arrow::internal::InvertBitmap(
+                                           ctx->memory_pool(), cond.buffers[1]->data(),
+                                           cond.offset, cond.length))
+      } else {
+        // out_buf = zeros
+        ARROW_ASSIGN_OR_RAISE(out_buf, ctx->AllocateBitmap(cond.length));
+      }
+    }
+    out->buffers[1] = std::move(out_buf);
+    return Status::OK();
+  }
+};
+
+template <typename Type>
+struct IfElseFunctor<Type, enable_if_null<Type>> {
+  template <typename T>
+  static inline Status ReturnCopy(const T& in, T* out) {
+    // Nothing preallocated, so we assign in into the output
+    *out = in;
+    return Status::OK();
+  }
+
+  // AAA
+  static Status Call(KernelContext* ctx, const ArrayData& cond, const ArrayData& left,
+                     const ArrayData& right, ArrayData* out) {
+    return ReturnCopy(left, out);
+  }
+
+  // ASA
+  static Status Call(KernelContext* ctx, const ArrayData& cond, const Scalar& left,
+                     const ArrayData& right, ArrayData* out) {
+    return ReturnCopy(right, out);
+  }
+
+  // AAS
+  static Status Call(KernelContext* ctx, const ArrayData& cond, const ArrayData& left,
+                     const Scalar& right, ArrayData* out) {
+    return ReturnCopy(left, out);
+  }
+
+  // ASS
+  static Status Call(KernelContext* ctx, const ArrayData& cond, const Scalar& left,
+                     const Scalar& right, ArrayData* out) {
+    return ReturnCopy(cond, out);
+  }
+};
+
+template <typename Type>
+struct ResolveIfElseExec {
+  static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    // cond is scalar
+    if (batch[0].is_scalar()) {
+      const auto& cond = batch[0].scalar_as<BooleanScalar>();
+      if (batch[1].is_scalar() && batch[2].is_scalar()) {
+        if (cond.is_valid) {
+          *out = cond.value ? batch[1].scalar() : batch[2].scalar();
+        } else {
+          *out = MakeNullScalar(batch[1].type());
+        }
+      } else {  // either left or right is an array. output is always an array
+        // output size is the size of the array arg
+        int64_t bcast_size = batch[1].is_array() ? batch[1].length() : batch[2].length();
+        if (cond.is_valid) {
+          const auto& valid_data = cond.value ? batch[1] : batch[2];
+          if (valid_data.is_array()) {
+            *out = valid_data;
+          } else {  // valid data is a scalar that needs to be broadcasted
+            ARROW_ASSIGN_OR_RAISE(*out,
+                                  MakeArrayFromScalar(*valid_data.scalar(), bcast_size,
+                                                      ctx->memory_pool()));
+          }
+        } else {  // cond is null. create null array
+          ARROW_ASSIGN_OR_RAISE(
+              *out, MakeArrayOfNull(batch[1].type(), bcast_size, ctx->memory_pool()))
+        }
+      }
+      return Status::OK();
+    }
+
```

Comment:
Instead of needing all those overloads for PropagateNulls, you could handle that here. Whether they are scalar or arrays, the salient point to null propagation is that each of these inputs may be entirely null (a null scalar or an array containing only nulls) or entirely valid (a valid scalar or an array with no nulls) or they have a null bitmap which needs to be visited. I'd write `util::optional<uint64_t> GetConstantValidityWord(const Datum&)` which you can apply to each input. If a a constant validity word can be created, you can use that directly in your visitor